### PR TITLE
[OUTDATED] Fix and rename `move_model_to_eval`

### PR DESCRIPTION
**Pull request by andrewor14**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #108801

Summary: This commit fixes two silent correctness problems with
the current implementation of `move_model_to_eval`:

(1) Previously the user had to manually call `eliminate_dead_code`
before calling `move_model_to_eval`, otherwise the dropout pattern
won't actually get eliminated. This is because subgraph rewriter
complains the match is not self-contained, and so silently does
not do the replacement.

(2) We wish to error when the user calls `model.train()` or
`model.eval()` on an exported model. This error is raised
correctly immediately after export today, but no longer raised
after the user calls prepare or convert.

We fix (1) by moving the `eliminate_dead_code` call into
`move_model_to_eval`, and fix (2) by ensuring the respective
errors are thrown after prepare and convert as well.

Additionally, this commit renames `move_model_to_eval` to
`move_exported_model_to_eval` to be more explicit.

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_disallow_eval_train
python test/test_quantization.py TestQuantizePT2E.test_move_exported_model_to_eval

Reviewers: jerryzh168, kimishpatel

Subscribers: jerryzh168, kimishpatel, supriyar

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

Differential Revision: [D49097293](https://our.internmc.facebook.com/intern/diff/D49097293)

### CODE REVIEW

File: torch/ao/quantization/pt2e/eval_utils.py [hunk 1]
```diff
@@ -17,6 +17,10 @@ def _replace_dropout_for_eval(m: torch.fx.GraphModule):
     # Avoid circular dependencies
     from .utils import get_aten_graph_module
 
+    # Needed to ensure subgraph matches are self-contained
+    m.graph.eliminate_dead_code()
+    m.recompile()
```
Comment (jerryzh168) [inline torch/ao/quantization/pt2e/eval_utils.py:22]: is there a test catching this?

**Review by jerryzh168 — APPROVED** [review]
LG, please add a test

File: torch/ao/quantization/pt2e/eval_utils.py (hunk 1 above)
Comment (andrewor14) [reply torch/ao/quantization/pt2e/eval_utils.py:22]: Yeah this is tested in the existing `test_move_exported_model_to_eval`

**Comment by andrewor14** [comment]
@andrewor14 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D49097293).

**Comment by andrewor14** [comment]
Closing in favor of #108891 due to export issues
